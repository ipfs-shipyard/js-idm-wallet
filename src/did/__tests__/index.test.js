import createIpid from '../methods/ipid';
import createDid from '../index';
import { mockDocument, mockIpid } from './mocks';

jest.mock('../methods/ipid', () => jest.fn(() => mockIpid));

beforeEach(() => {
    jest.clearAllMocks();
});

it('should have all specification methods', async () => {
    const did = createDid();

    expect(typeof did.resolve).toEqual('function');
    expect(typeof did.create).toEqual('function');
    expect(typeof did.update).toEqual('function');
    expect(typeof did.isPublicKeyValid).toEqual('function');
    expect(typeof did.getMethods).toEqual('function');
});

describe('getMethods', () => {
    it('should return all available methods and its information', async () => {
        const did = createDid();

        expect(did.getMethods()).toMatchSnapshot();
    });
});

describe('resolve', () => {
    it('should resolve successfully', async () => {
        const did = createDid();
        const document = await did.resolve('did:ipid:foo');

        expect(mockIpid.resolve).toHaveBeenCalledTimes(1);
        expect(mockIpid.resolve).toHaveBeenCalledWith('did:ipid:foo');
        expect(document).toEqual(mockDocument);
    });

    it('should fail if invalid did', async () => {
        expect.assertions(2);

        const did = createDid();

        try {
            await did.resolve('did#abcdef');
        } catch (err) {
            expect(err.message).toEqual('Invalid DID: did#abcdef');
            expect(err.code).toEqual('INVALID_DID');
        }
    });

    it('should fail if method is not supported', async () => {
        expect.assertions(2);

        const did = createDid();

        try {
            await did.resolve('did:fake:abcdef');
        } catch (err) {
            expect(err.message).toEqual('Did method `fake` is not supported');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD');
        }
    });

    it('should fail if method does not support purpose', async () => {
        const mockIpidOnce = { ...mockIpid, resolve: undefined };

        createIpid.mockImplementationOnce(jest.fn(() => mockIpidOnce));

        expect.assertions(2);

        const did = createDid();

        try {
            await did.resolve('did:ipid:abcdef');
        } catch (err) {
            expect(err.message).toEqual('Purpose `resolve` is not currently supported for `ipid`');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD_PURPOSE');
        }
    });
});

describe('create', () => {
    it('should create successfully', async () => {
        const mockParams = { privateKey: 'mockPrivateKey', foo: 'bar' };
        const mockOperations = jest.fn();

        const did = createDid();
        const document = await did.create('ipid', mockParams, mockOperations);

        expect(mockIpid.create).toHaveBeenCalledTimes(1);
        expect(mockIpid.create).toHaveBeenCalledWith(mockParams, mockOperations);
        expect(document).toEqual(mockDocument);
    });

    it('should fail if method is not supported', async () => {
        expect.assertions(2);

        const did = createDid();

        try {
            await did.create('fake');
        } catch (err) {
            expect(err.message).toEqual('Did method `fake` is not supported');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD');
        }
    });

    it('should fail if method does not support purpose', async () => {
        const mockIpidOnce = { ...mockIpid, create: undefined };

        createIpid.mockImplementationOnce(jest.fn(() => mockIpidOnce));

        expect.assertions(2);

        const did = createDid();

        try {
            await did.create('ipid');
        } catch (err) {
            expect(err.message).toEqual('Purpose `create` is not currently supported for `ipid`');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD_PURPOSE');
        }
    });
});

describe('update', () => {
    it('should create successfully', async () => {
        const mockParams = { privateKey: 'mockPrivateKey', foo: 'bar' };
        const mockOperations = jest.fn();

        const did = createDid();
        const document = await did.update('did:ipid:abcdef', mockParams, mockOperations);

        expect(mockIpid.update).toHaveBeenCalledTimes(1);
        expect(mockIpid.update).toHaveBeenCalledWith('did:ipid:abcdef', mockParams, mockOperations);
        expect(document).toEqual(mockDocument);
    });

    it('should fail if invalid did', async () => {
        expect.assertions(2);

        const did = createDid();

        try {
            await did.update('did#abcdef');
        } catch (err) {
            expect(err.message).toEqual('Invalid DID: did#abcdef');
            expect(err.code).toEqual('INVALID_DID');
        }
    });

    it('should fail if method is not supported', async () => {
        expect.assertions(2);

        const did = createDid();

        try {
            await did.update('did:fake:abcdef');
        } catch (err) {
            expect(err.message).toEqual('Did method `fake` is not supported');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD');
        }
    });

    it('should fail if method does not support purpose', async () => {
        const mockIpidOnce = { ...mockIpid, update: undefined };

        createIpid.mockImplementationOnce(jest.fn(() => mockIpidOnce));

        expect.assertions(2);

        const did = createDid();

        try {
            await did.update('did:ipid:abcdef');
        } catch (err) {
            expect(err.message).toEqual('Purpose `update` is not currently supported for `ipid`');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD_PURPOSE');
        }
    });
});

describe('isPublicKeyValid', () => {
    it('should be successful if public key in document', async () => {
        const mockOptions = { foo: 'bar' };

        const did = createDid();
        const isValid = await did.isPublicKeyValid('did:ipid:abcdef', 'did:ipid:abcdef#123', mockOptions);

        expect(isValid).toBeTruthy();
        expect(mockIpid.isPublicKeyValid).toHaveBeenCalledTimes(1);
        expect(mockIpid.isPublicKeyValid).toHaveBeenCalledWith('did:ipid:abcdef', 'did:ipid:abcdef#123', mockOptions);
    });

    it('should fail if invalid did', async () => {
        expect.assertions(2);

        const did = createDid();

        try {
            await did.isPublicKeyValid('did#abcdef');
        } catch (err) {
            expect(err.message).toEqual('Invalid DID: did#abcdef');
            expect(err.code).toEqual('INVALID_DID');
        }
    });

    it('should fail if method is not supported', async () => {
        expect.assertions(2);

        const did = createDid();

        try {
            await did.isPublicKeyValid('did:fake:abcdef');
        } catch (err) {
            expect(err.message).toEqual('Did method `fake` is not supported');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD');
        }
    });

    it('should fail if method does not support purpose', async () => {
        const mockIpidOnce = { ...mockIpid, isPublicKeyValid: undefined };

        createIpid.mockImplementationOnce(jest.fn(() => mockIpidOnce));

        expect.assertions(2);

        const did = createDid();

        try {
            await did.isPublicKeyValid('did:ipid:abcdef');
        } catch (err) {
            expect(err.message).toEqual('Purpose `isPublicKeyValid` is not currently supported for `ipid`');
            expect(err.code).toEqual('UNSUPPORTED_DID_METHOD_PURPOSE');
        }
    });
});
