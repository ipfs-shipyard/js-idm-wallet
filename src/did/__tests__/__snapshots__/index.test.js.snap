// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMethods should return all available methods and its information 1`] = `
Array [
  Object {
    "description": "The Interplanetary Identifiers DID method",
    "method": "ipid",
  },
]
`;
